syntax = "proto3";

package transaction_verification;

service TransactionVerification {
    rpc VerifyTransaction(TransactionRequest) returns (TransactionResponse);
    rpc ClearOrderData(ClearOrderRequest) returns (ClearOrderResponse); // 新增清理订单方法
}

message TransactionRequest {
    string order_id = 1;
    string user_id = 2;
    float amount = 3;
    CreditCard credit_card = 4;
    string discount_code = 5;
    BillingAddress billing_address = 6;
    string shipping_method = 7;
    string user_comment = 8;
    bool gift_wrapping = 9;
    repeated int32 vector_clock = 10; // 新增向量时钟字段
}

message CreditCard {
    string number = 1;
    string expiration_date = 2;
    string cvv = 3;
}

message BillingAddress {
    string street = 1;
    string city = 2;
    string state = 3;
    string zip = 4;
    string country = 5;
}

message TransactionResponse {
    bool approved = 1;
    string reason = 2;
    TransactionDetails details = 3;
    repeated int32 vector_clock = 4; // 传递向量时钟
}

message TransactionDetails {
    string order_id = 1;
    string status = 2;
    BillingAddress billing_address = 3;
    string discount_code = 4;
    float charged_amount = 5;
    string shipping_method = 6;
    string user_comment = 7;
    bool gift_wrapping = 8;
}

message ClearOrderRequest {
    string order_id = 1;
    repeated int32 vector_clock = 2; // 传递当前的向量时钟
}

message ClearOrderResponse {
    bool success = 1;
}
